packages/astro/src/server/middleware.ts

```diff
@@ -93,11 +95,30 @@ export const handleRequest: (options?: MiddlewareOptions) => MiddlewareResponseH
           },
         },
         async span => {
-          const res = await next();
-          if (span && res.status) {
-            span.setHttpStatus(res.status);
+          const originalResponse = await next();
+          if (span && originalResponse.status) {
+            span.setHttpStatus(originalResponse.status);
+          }
+
+          const hub = getCurrentHub();
+          const client = hub.getClient();
+          const contentType = originalResponse.headers.get('content-type');
+
+          const isPageloadRequest = contentType && contentType.startsWith('text/html');
+          if (!isPageloadRequest || !client) {
+            return originalResponse;
           }
-          return res;
+
+          const html = await originalResponse.text();
```

Comment:
m: Honestly I am a little worried here for multiple reasons:
- I don't understand the Request/Response APIs enough yet but my worry is that this will consume the repsonse stream and somehow make it unavailable for anything that tries to access it.
- I believe with our approach here we will definitely prevent streamed responses from being streamed and just completely block right? This _could_ hurt TTFB.
- I don't know how to feel about the heuristic to find the head tag here 😂. Using an HTML parser to properly find it is probably overkill?
- Let's make absolutely sure this doesn't open any XSS vectors. Maybe we run it through a regexp before writing to the HTML.

Reply:
> I don't understand the Request/Response APIs enough yet but my worry is that this will consume the repsonse stream and somehow make it unavailable for anything that tries to access it.
I believe with our approach here we will definitely prevent streamed responses from being streamed and just completely block right? This could hurt TTFB.

It's a good point. Astro [uses streaming](https://docs.astro.build/en/guides/server-side-rendering/#streaming) but they also show a [similar example](https://docs.astro.build/en/guides/middleware/#example-redacting-sensitive-information) for modifying HTML in their middleware docs. I'm gonna ask the Astro folks if there are performance concerns with doing this.

 > I don't know how to feel about the heuristic to find the head tag here 😂. Using an HTML parser to properly find it is probably overkill?

Agreed, it's super basic but we use the identical approach in SvelteKit and so far it worked decently well. Using a parser is for sure a performance concern. My pragmatic take would be to go with this and improve the <head> lookup logic once we discover problems here. 

> Let's make absolutely sure this doesn't open any XSS vectors. 

Generally, I agree but I'm not entirely sure about the XSS vector. I guess for something malicious to end up in here, the SDK's options (release, environment) or the transaction data (name, ids) would need to be somehow modified beforehand. Do you see any obvious ways how this could happen?

> Maybe we run it through a regexp before writing to the HTML.
I think this works decently well for `sentry-trace` content but baggage is a little more arbitrary. Lemme try to come up with something.

Reply:
> Astro [uses streaming](https://docs.astro.build/en/guides/server-side-rendering/#streaming) but they also show a [similar example](https://docs.astro.build/en/guides/middleware/#example-redacting-sensitive-information) for modifying HTML

Alright if Astro is recommending this it's probably fine.

> My pragmatic take would be to go with this and improve the lookup logic once we discover problems here.

Sounds good to me.

> Do you see any obvious ways how this could happen?

No obvious ways, but if we can come up with a simple way to ensure the string cant be escaped it would be great.

Reply:
@lforst I added a baggage regex check in [272cbbd](https://github.com/getsentry/sentry-javascript/pull/9483/commits/272cbbdde7b369c2bcce132608903c707f15c21a)

for sentry-trace, we already had a regex 

We'll only serialize the baggage/sentry-trace content if the regexes match.